preview/qt_preview.cpp

```diff
@@ -0,0 +1,175 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (C) 2021, Raspberry Pi (Trading) Ltd.
+ *
+ * null_preview.cpp - dummy "show nothing" preview window.
+ */
+
+#include <condition_variable>
+#include <iostream>
+#include <mutex>
+#include <thread>
+
+#include <QApplication>
+#include <QImage>
+#include <QMainWindow>
+#include <QPaintEvent>
+#include <QPainter>
+#include <QWidget>
+
+#include "preview.hpp"
+
+class MyMainWindow : public QMainWindow
+{
+public:
+	MyMainWindow() : QMainWindow() {}
+protected:
+	// This prevents a crash, you could connect it up properly if you wanted.
+	void closeEvent(QCloseEvent *event) override { event->ignore(); }
+};
+
+class MyWidget : public QWidget
+{
+public:
+	MyWidget(QWidget *parent, int w, int h) : QWidget(parent), size(w, h)
+	{
+		image = QImage(size, QImage::Format_RGB888);
+		image.fill(0);
+	}
+	QSize size;
+	QImage image;
+protected:
+	void paintEvent(QPaintEvent *) override
+	{
+		QPainter painter(this);
+		painter.drawImage(rect(), image, image.rect());
+	}
+	QSize sizeHint() const override { return size; }
+};
+
+class QtPreview : public Preview
+{
+public:
+	QtPreview(Options const *options) : Preview(options)
+	{
+		window_width_ = options->preview_width;
+		window_height_ = options->preview_height;
+		// This preview window is expensive, so make it small by default.
+		if (window_width_ == 0 || window_height_ == 0)
+			window_width_ = 512, window_height_ = 384;
+		thread_ = std::thread(&QtPreview::threadFunc, this, options);
+		std::unique_lock lock(mutex_);
+		while (!pane_)
+			cond_var_.wait(lock);
+		if (options->verbose)
+			std::cout << "Made Qt preview" << std::endl;
+	}
+	~QtPreview()
+	{
+		application_->exit();
+		thread_.join();
+	}
+	void SetInfoText(const std::string &text) override { main_window_->setWindowTitle(QString::fromStdString(text)); }
+	virtual void Show(int fd, void *mem, size_t size, int width, int height, int stride) override
+	{
+		// Cache the x sampling locations for speed. This is a quick nearest neighbour resize.
+		if (last_image_width_ != width)
+		{
+			last_image_width_ = width;
+			x_locations_.resize(window_width_);
+			for (int i = 0; i < window_width_; i++)
+				x_locations_[i] = (i * (width - 1) + (window_width_ - 1) / 2) / (window_width_ - 1);
+		}
+
+		uint8_t *Y_start = (uint8_t *)mem;
+		uint8_t *U_start = Y_start + stride * height;
+		int uv_size = (stride / 2) * (height / 2);
+		uint8_t *dest = pane_->image.bits();
+
+		// Possibly this should be locked in case a repaint is happening? In practice the risk
+		// is only that there might be some tearing, so I don't think we worry. We could speed
+		// it up by getting the ISP to supply RGB, but I'm not sure I want to handle that extra
+		// possibility in our main application code, so we'll put up with the slow conversion.
+		for (int y = 0; y < window_height_; y++)
```

Comment:
Perhaps also an assert to make sure we are coming into here with a YUV420 format image?

Reply:
All the preview windows only support YUV420 so we probably want to check back in the LibcameraApp::previewThread, but doing so would clearly be sensible.